File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> preSumMap;
        int sum = 0, cnt = 0;
        for(int i = 0; i < nums.size(); i++){
            sum += nums[i];
            if(sum == k){
                cnt++;
            }
            int rem = sum - k;
            if(preSumMap.find(rem) != preSumMap.end()){
                cnt += preSumMap[rem];
            }
            if(preSumMap.find(sum) == preSumMap.end()){
                preSumMap[sum] = 1; // counting the frequency
            }
            else preSumMap[sum]++;
        }
        return cnt;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer,Integer> mpp = new HashMap<>();
        mpp.put(0,1);
        int preSum = 0, cnt = 0;
        for(int i = 0; i < nums.length; i++){
            preSum += nums[i];
            int rem = preSum - k;
            if(mpp.containsKey(rem))    cnt += mpp.get(rem);
            mpp.put(preSum, mpp.getOrDefault(preSum, 0) + 1);
        }
        return cnt;
    }
}